synapse/tests/test_cortex.py

```diff
@@ -101,6 +101,70 @@ async def test_cortex_divert(self):
             self.len(2, [n for n in nodes if n.ndef[0] == 'inet:fqdn'])
             self.len(2, await core.nodes('ou:org +#faz'))
 
+            # functions that don't return a generator
+            storm = '''
+            function x() {
+                $lst = $lib.list()
+                [ ou:org=* ou:org=* +#cat ]
+                $lst.append($node)
+                fini { return($lst) }
+            }
+
+            divert $lib.true $x()
+            '''
+            await self.asyncraises(s_exc.BadArg, core.nodes(storm))
+            self.len(2, await core.nodes('ou:org +#cat'))
+
+            storm = '''
+            function x() {
+                $lst = $lib.list()
+                [ ou:org=* ou:org=* +#dog ]
+                $lst.append($node)
+                fini { return($lst) }
+            }
+
+            [ inet:fqdn=vertex.link inet:fqdn=woot.com ]
+
+            divert $lib.true $x()
+            '''
+            await self.asyncraises(s_exc.BadArg, core.nodes(storm))
+            self.len(4, await core.nodes('ou:org +#dog'))
+
+            # functions that return a generator
+            storm = '''
+            function y() {
+                [ ou:org=* ou:org=* +#cow ]
+            }
+            function x() {
+                $lib.print(heythere)
+                return($y())
+            }
+
+            divert $lib.true $x()
+            '''
+            mesgs = await core.stormlist(storm)
+            self.len(1, [m for m in mesgs if (m[0] == 'print' and m[1]['mesg'] == 'heythere')])
+            self.len(2, [m for m in mesgs if (m[0] == 'node' and m[1][0][0] == 'ou:org')])
+            self.len(2, await core.nodes('ou:org +#cow'))
+
+            storm = '''
+            function y() {
+                [ ou:org=* ou:org=* +#camel ]
+            }
+            function x() {
+                $lib.print(heythere)
+                return($y())
+            }
+
+            [ inet:fqdn=vertex.link inet:fqdn=woot.com ]
+
+            divert $lib.false $x()
+            '''
+            mesgs = await core.stormlist(storm)
+            self.len(3, [m for m in mesgs if (m[0] == 'print' and m[1]['mesg'] == 'heythere')])
```

Comment:
Note that this works since we are returning a generator, but x() still gets invoked multiple times.  We would have to reach into the query to detect if the function has a return if we wanted to disallow this.